<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Identification of CRISPR offtargets by BreakTag • breakinspectoR</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Identification of CRISPR offtargets by BreakTag">
<meta name="description" content="Companion package to the BreakTag protocol for the identification of CRISPR offtargets.">
<meta property="og:description" content="Companion package to the BreakTag protocol for the identification of CRISPR offtargets.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">breakinspectoR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="breakinspector">breakinspectoR<a class="anchor" aria-label="anchor" href="#breakinspector"></a>
</h1></div>
<p>A companion R package to the BreakTag protocol for the identification of CRISPR offtargets. breakinspectoR is an R package which performs a guided search toward putative on-/off-targets.</p>
<div class="section level2">
<h2 id="overall-description">Overall description<a class="anchor" aria-label="anchor" href="#overall-description"></a>
</h2>
<div class="float">
<img src="assets/breakinspectoR_wf.png" alt="breakinspectoR workflow"><div class="figcaption">breakinspectoR workflow</div>
</div>
<div class="section level3">
<h3 id="initial-preprocessing">Initial preprocessing<a class="anchor" aria-label="anchor" href="#initial-preprocessing"></a>
</h3>
<p>Initial preprocessing is typically done in a linux cluster using the <a href="https://github.com/roukoslab/breaktag" class="external-link">Breaktag pipeline</a>. It includes the following steps:</p>
<ol style="list-style-type: decimal">
<li>scanning for reads (single- or paired-end) containing the expected 8-nt UMI followed by the 8-nt sample barcode in the 5’ end of read 1.</li>
<li>alignment of reads to reference genome with BWA, with a seed length of 19 and default scoring/penalty values for mismatches, gaps and read clipping.</li>
<li>reads mapped with a minimum quality score Q (defaults to Q=60) are retained.</li>
<li>close spatial consecutive reads within a window of 30 nucleotides and UMI differing with up to 2 mismatches are considered PCR duplicates and only one is kept.</li>
</ol>
<p>The resulting reads are aggregated per position and reported as a BED file.</p>
</div>
<div class="section level3">
<h3 id="breakinspector-analysis">breakinspectoR analysis<a class="anchor" aria-label="anchor" href="#breakinspector-analysis"></a>
</h3>
<p>The analysis path consists of the following steps:</p>
<ol style="list-style-type: decimal">
<li>from the previously generated BED file, identify stacks of read ends as the candidate loci of being CRISPR-edited.</li>
<li>obtain the sequence context of the candidate loci, and keep only those which are at most <code>N</code> nucleotides upstream from a PAM and contain up to M mismatches with regard to the gDNA guide sequence. Defaults are the canonical N=3, M=7, PAM=“NGG”.</li>
<li>count number of reads (== signal or DSBreaks) in the targeted and the non-targeted control library.</li>
<li>test if the enrichment of reads we see in the targeted library is significant compared to the non-targeted control library. Here breakinspectoR will perform a binomial test with the following criteria:
<ol style="list-style-type: decimal">
<li>preconditions:
<ol style="list-style-type: decimal">
<li>A = number of breaks within the region in the target library</li>
<li>B = total number of breaks in the whole target library</li>
<li>C = number of breaks within the region in the nontarget library</li>
<li>D = total number of breaks in the whole nontarget library</li>
</ol>
</li>
<li>the binomial model for calculating the p-Value is:
<ol style="list-style-type: decimal">
<li>number of trials = A+1</li>
<li>number of successes = B+1</li>
<li>estimated success probability in each trial = (C+1) / (D+1)</li>
</ol>
</li>
<li>the enrichment is then calculated as: ((A+1) / (B+1)) / ((C+1) / (D+1))</li>
<li>note that we add a pseudocount to avoid dividing by 0</li>
<li>q-values and local False Discovery Rate values are estimated for FDR control.</li>
</ol>
</li>
<li>additionally, breakinspectoR implements a complimentary [and elaborated] method to estimate the false discovery of targets. To summarize, breakinspectoR reshuffles the signal in the target library using several multinomially distributed random number vectors sampled with equal probabilities to the signal in the originally detected offtargets. Then, breakinspectoR analysis is done in the reshuffled target library vs. the non-target library, and an FDR is estimated comparing the signal of each offtarget called in the original target library to the targets detected in the reshuffled target library (where no targets were expected to be called).</li>
<li>breakinspectoR includes several handy visualizations to further analyze and summarize the on-/off-targets detected. Some of these functions include the analysis of fidelity of the gDNA, sequence composition of target regions, frequency of mismatches per position of the protospacer, or the genomic distribution of the targeted regions.</li>
</ol>
</div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>Open R and install directly from Github with <code>devtools</code> (install the package <code>devtools</code> if you haven’t, yet):</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"roukoslab/breakinspectoR"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="example-usage">Example usage<a class="anchor" aria-label="anchor" href="#example-usage"></a>
</h2>
<p>This is a simple example using the demo data for human chr6 included with the package. The experiment identifies offtargets generated by the VEGFA site 2 sgRNA with CRISPR/Cas9.</p>
<p>Call the breakinspectoR analysis to find offtargets enriched in the targeted library compared to the non-targeted. We’ll stick to the default 7 mismatches allowed to the guide, with the expected cut site 3 bp away from the PAM (Cas9).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">target_file</span>     <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/vegfa.chr6.bed.gz"</span>, package<span class="op">=</span><span class="st">"breakinspectoR"</span><span class="op">)</span></span>
<span><span class="va">non_target_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/nontarget.chr6.bed.gz"</span>, package<span class="op">=</span><span class="st">"breakinspectoR"</span><span class="op">)</span></span>
<span><span class="va">guide</span>           <span class="op">&lt;-</span> <span class="st">"GACCCCCTCCACCCCGCCTC"</span></span>
<span><span class="va">PAM</span>             <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>canonical<span class="op">=</span><span class="st">"NGG"</span>, <span class="st">"NAG"</span><span class="op">)</span></span>
<span><span class="va">bsgenome</span>        <span class="op">&lt;-</span> <span class="st">"BSgenome.Hsapiens.UCSC.hg38"</span></span>
<span></span>
<span><span class="va">offtargets</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/breakinspectoR.html">breakinspectoR</a></span><span class="op">(</span></span>
<span>    target        <span class="op">=</span><span class="va">target_file</span>,</span>
<span>    nontarget     <span class="op">=</span><span class="va">non_target_file</span>,</span>
<span>    guide         <span class="op">=</span><span class="va">guide</span>,</span>
<span>    PAM           <span class="op">=</span><span class="va">PAM</span>,</span>
<span>    bsgenome      <span class="op">=</span><span class="va">bsgenome</span>,</span>
<span>    cutsiteFromPAM<span class="op">=</span><span class="fl">3</span>,</span>
<span>    verbose       <span class="op">=</span><span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The analysis will take few seconds to run. Afterwards we have a comprehensive table with few hundred enriched offtarget loci, which we can summarize using the accompanying plotting functions:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_position_cutsite.html">plot_position_cutsite</a></span><span class="op">(</span><span class="va">offtargets</span>, guide<span class="op">=</span><span class="va">guide</span>, pam<span class="op">=</span><span class="va">PAM</span><span class="op">[</span><span class="st">"canonical"</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/plot_mismatch_freq.html">plot_mismatch_freq</a></span><span class="op">(</span><span class="va">offtargets</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/plot_offtargets_by_pam.html">plot_offtargets_by_pam</a></span><span class="op">(</span><span class="va">offtargets</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/plot_sequence_composition.html">plot_sequence_composition</a></span><span class="op">(</span><span class="va">offtargets</span>, guide<span class="op">=</span><span class="va">guide</span>, pam<span class="op">=</span><span class="va">PAM</span><span class="op">[</span><span class="st">"canonical"</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/plot_guide_fidelity.html">plot_guide_fidelity</a></span><span class="op">(</span><span class="va">offtargets</span>, guide<span class="op">=</span><span class="va">guide</span>, pam<span class="op">=</span><span class="va">PAM</span><span class="op">[</span><span class="st">"canonical"</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/plot_genomic_position.html">plot_genomic_position</a></span><span class="op">(</span><span class="va">offtargets</span>, bsgenome<span class="op">=</span><span class="va">bsgenome</span>,</span>
<span>                      target<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/vegfa.chr6.bed.gz"</span>, package<span class="op">=</span><span class="st">"breakinspectoR"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/plot_pam_logo.html">plot_pam_logo</a></span><span class="op">(</span><span class="va">offtargets</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/plot_pam_composition.html">plot_pam_composition</a></span><span class="op">(</span><span class="va">offtargets</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/manhattan_plot.html">manhattan_plot</a></span><span class="op">(</span><span class="va">offtargets</span>, bsgenome<span class="op">=</span><span class="va">bsgenome</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># simulate 2 different experiments by picking 25 random offtargets</span></span>
<span><span class="va">exp1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">offtargets</span><span class="op">[</span><span class="va">offtargets</span><span class="op">$</span><span class="va">mismatches</span> <span class="op">==</span> <span class="fl">0</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">offtargets</span>, <span class="fl">25</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">exp2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">offtargets</span><span class="op">[</span><span class="va">offtargets</span><span class="op">$</span><span class="va">mismatches</span> <span class="op">==</span> <span class="fl">0</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">offtargets</span>, <span class="fl">25</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/plot_relative_activity.html">plot_relative_activity</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>a<span class="op">=</span><span class="va">exp1</span>, b<span class="op">=</span><span class="va">exp2</span><span class="op">)</span>, what<span class="op">=</span><span class="st">"all"</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/plot_specificity.html">plot_specificity</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>a<span class="op">=</span><span class="va">exp1</span>, b<span class="op">=</span><span class="va">exp2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/plot_blunt_rate_density.html">plot_blunt_rate_density</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>a<span class="op">=</span><span class="va">exp1</span>, b<span class="op">=</span><span class="va">exp2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="scission-profile-analysis">Scission profile analysis<a class="anchor" aria-label="anchor" href="#scission-profile-analysis"></a>
</h3>
<p>One unique feature of BreakTag is that it provides the structure of a DSB at single base resolution. BreakinspectoR implements a method and several visualizations to analyze DSB end structure as determined by BreakTag.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">offtargets.scission_profile</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/scission_profile_analysis.html">scission_profile_analysis</a></span><span class="op">(</span></span>
<span>  x        <span class="op">=</span><span class="va">offtargets</span>,</span>
<span>  target   <span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/vegfa.chr6.bed.gz"</span>, package<span class="op">=</span><span class="st">"breakinspectoR"</span><span class="op">)</span>,</span>
<span>  nontarget<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/nontarget.chr6.bed.gz"</span>, package<span class="op">=</span><span class="st">"breakinspectoR"</span><span class="op">)</span>,</span>
<span>  bsgenome <span class="op">=</span><span class="st">"BSgenome.Hsapiens.UCSC.hg38"</span><span class="op">)</span></span></code></pre></div>
<p>Along with several useful visualizations:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># simulate 2 different experiments by picking 25 random offtargets</span></span>
<span><span class="va">exp1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">offtargets.scission_profile</span>, <span class="fl">25</span><span class="op">)</span></span>
<span><span class="va">exp2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">offtargets.scission_profile</span>, <span class="fl">25</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="reference/plot_blunt_rate_density.html">plot_blunt_rate_density</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>a<span class="op">=</span><span class="va">exp1</span>, b<span class="op">=</span><span class="va">exp2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/plot_overhang_size.html">plot_overhang_size</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>a<span class="op">=</span><span class="va">exp1</span>, b<span class="op">=</span><span class="va">exp2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/plot_scission_profile.html">plot_scission_profile</a></span><span class="op">(</span><span class="va">offtargets.scission_profile</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="input-files">Input files<a class="anchor" aria-label="anchor" href="#input-files"></a>
</h2>
<p>Input BED files describing coordinates and number of DSB are expected for the “target” and “non-target” libraries. It’s possible to run breakinspectoR without the non-target library, nevertheless it is advised to include such experiment to calculate a p-value and control the false discovery rate.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="ex">chr6</span>  148074  148075  .  2  +</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="ex">chr6</span>  148093  148094  .  1  <span class="at">-</span></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="ex">chr6</span>  148240  148241  .  1  <span class="at">-</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="ex">chr6</span>  148503  148504  .  1  +</span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="ex">chr6</span>  148636  148637  .  1  <span class="at">-</span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a><span class="ex">chr6</span>  148697  148698  .  1  <span class="at">-</span></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a><span class="ex">chr6</span>  149009  149010  .  1  <span class="at">-</span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a><span class="ex">chr6</span>  149363  149364  .  1  +</span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a><span class="ex">chr6</span>  150252  150253  .  2  +</span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a><span class="ex">chr6</span>  150263  150264  .  1  +</span></code></pre></div>
<p>These files are typically created with the <a href="https://github.com/roukoslab/breaktag" class="external-link">Breaktag pipeline</a>, although any conforming BED file is accepted in breakinspectoR.</p>
</div>
<div class="section level2">
<h2 id="btmotif">BTmotif<a class="anchor" aria-label="anchor" href="#btmotif"></a>
</h2>
<p>Additionally, you may want to run the companion shiny app to derive Cas9 sequence determinants from BreakInspectoR output.</p>
<p>It uses XGBoost and the provided sequence (usually, a protospacer) to predict which nucleotides and positions are important to predict any numerical outcome (eg. the blunt rate, Cas9 activity, etc.).</p>
<div class="float">
<img src="assets/BTmotif.png" alt="BTmotif"><div class="figcaption">BTmotif</div>
</div>
<div class="section level3">
<h3 id="dependencies">Dependencies<a class="anchor" aria-label="anchor" href="#dependencies"></a>
</h3>
<p>The main dependency is <a href="https://h2o.ai/" class="external-link">H2O</a>, which can be installed from CRAN. The app has been tested with H2O version 3.36.1.2.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"h2o"</span><span class="op">)</span></span></code></pre></div>
<p>Test the H2O installation with:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/h2oai/h2o-3" class="external-link">h2o</a></span><span class="op">)</span></span>
<span><span class="va">localH2O</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/h2o/man/h2o.init.html" class="external-link">h2o.init</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/demo.html" class="external-link">demo</a></span><span class="op">(</span><span class="va">h2o.kmeans</span><span class="op">)</span></span></code></pre></div>
<p>You’ll need a couple of packages to run the web app:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"shiny"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"shinydashboard"</span><span class="op">)</span></span></code></pre></div>
<p>To generate the motifs, you’ll also need <code>ggplot2</code> and <code>ggseqlogo</code>:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"ggplot2"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"omarwagih/ggseqlogo"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="run">Run<a class="anchor" aria-label="anchor" href="#run"></a>
</h3>
<p>Open the web app in your R console:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">breakinspectoR</span><span class="fu">::</span><span class="fu"><a href="reference/shiny_BTmotif.html">shiny_BTmotif</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Paste a table of targets and click on <code>Go!</code>. Or check the <code>Example</code> data. The list can actually be a table with &lt;tab&gt; or &lt;comma&gt; separated fields. The columns are expected to be in this order: protospacer_sequence | blunt_rate.</p>
</div>
</div>
<div class="section level2">
<h2 id="bluntpred">bluntPred<a class="anchor" aria-label="anchor" href="#bluntpred"></a>
</h2>
<p>For your set of gRNAs, you may want to run the prediction of blunt rates of Streptococcus pyogenes Cas9 (SpCas9) using the XGBoost model trained with HiPlex1 data.</p>
<div class="float">
<img src="assets/bluntPred.png" alt="bluntPred"><div class="figcaption">bluntPred</div>
</div>
<div class="section level3">
<h3 id="dependencies-1">Dependencies<a class="anchor" aria-label="anchor" href="#dependencies-1"></a>
</h3>
<p>The main dependency is <a href="https://h2o.ai/" class="external-link">H2O</a>.</p>
<p>Remove any previously installed H2O packages for R.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="st">"package:h2o"</span> <span class="op"><a href="https://rdrr.io/pkg/h2o/man/h2o.match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/search.html" class="external-link">search</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span> <span class="kw"><a href="https://rdrr.io/r/base/detach.html" class="external-link">detach</a></span><span class="op">(</span><span class="st">"package:h2o"</span>, unload<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="op">}</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="st">"h2o"</span> <span class="op"><a href="https://rdrr.io/pkg/h2o/man/h2o.match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span> <span class="fu"><a href="https://rdrr.io/r/utils/remove.packages.html" class="external-link">remove.packages</a></span><span class="op">(</span><span class="st">"h2o"</span><span class="op">)</span> <span class="op">}</span></span></code></pre></div>
<p>Download packages that H2O depends on.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RCurl"</span>,<span class="st">"jsonlite"</span>, <span class="st">"devtools"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Download and install the H2O package for R. The models were trained on H2O version 3.36.1.2, therefore specifically install this version.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"h2o"</span>, type<span class="op">=</span><span class="st">"source"</span>, repos<span class="op">=</span><span class="st">"http://h2o-release.s3.amazonaws.com/h2o/rel-zumbo/2/R"</span><span class="op">)</span></span></code></pre></div>
<p>Test the H2O installation with:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/h2oai/h2o-3" class="external-link">h2o</a></span><span class="op">)</span></span>
<span><span class="va">localH2O</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/h2o/man/h2o.init.html" class="external-link">h2o.init</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/demo.html" class="external-link">demo</a></span><span class="op">(</span><span class="va">h2o.kmeans</span><span class="op">)</span></span></code></pre></div>
<p>Now, it’s all set to install the package.</p>
<p>The package resides in GitHub only. You will probably need <code>devtools</code> for that (<code>install.packages("devtools")</code>).</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"roukoslab/bluntPred"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="run-1">Run<a class="anchor" aria-label="anchor" href="#run-1"></a>
</h3>
<p>Open the web app in your R console:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">bluntPred</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/bluntPred/man/shiny_bluntPred.html" class="external-link">shiny_bluntPred</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Paste a list of gRNAs targets and click on <code>Predict</code>. The list can actually be a table with &lt;tab&gt; or &lt;comma&gt; separated fields. The gRNA sequence is expected to be in the <em>first</em> column.</p>
<p>NOTE: Only the seed portion of the protospacer (this is, the last 10 nucloetides of the target sequence) are used for the prediction in this model.</p>
</div>
</div>
<div class="section level2">
<h2 id="cite">Cite<a class="anchor" aria-label="anchor" href="#cite"></a>
</h2>
<p>If you find this tool useful and use it in your research, please cite our publication:</p>
<p>Longo, Sayols et al., Linking CRISPR–Cas9 double-strand break profiles to gene editing precision with BreakTag. Nat. Biotechnol. 2024. DOI: <a href="https://doi.org/10.1038/s41587-024-02238-8" class="external-link uri">https://doi.org/10.1038/s41587-024-02238-8</a></p>
</div>
</div>
  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing breakinspectoR</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Sergi Sayols <br><small class="roles"> Author, maintainer </small>  </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Sergi Sayols.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
