<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>scission_profile_analysis — scission_profile_analysis • breakinspectoR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="scission_profile_analysis — scission_profile_analysis"><meta name="description" content="Analyses the scission profile, whether the Cas9 protein cut blunt or
staggered around the cut site. A significant p-value means that most likely
Cas9 performed a staggered cut."><meta property="og:description" content="Analyses the scission profile, whether the Cas9 protein cut blunt or
staggered around the cut site. A significant p-value means that most likely
Cas9 performed a staggered cut."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">breakinspectoR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>scission_profile_analysis</h1>

      <div class="d-none name"><code>scission_profile_analysis.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Analyses the scission profile, whether the Cas9 protein cut blunt or
staggered around the cut site. A significant p-value means that most likely
Cas9 performed a staggered cut.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">scission_profile_analysis</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">target</span>,</span>
<span>  nontarget <span class="op">=</span> <span class="fu">GRanges</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  region <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  standard_chromosomes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  bsgenome <span class="op">=</span> <span class="st">"BSgenome.Hsapiens.UCSC.hg38"</span>,</span>
<span>  qval_cutoff <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  cores <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"mc.cores"</span>, <span class="fl">2L</span><span class="op">)</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>A GRanges object containing the results of a breakinspectoR
analysis.</p></dd>


<dt id="arg-target">target<a class="anchor" aria-label="anchor" href="#arg-target"></a></dt>
<dd><p>GRanges object with coordinates of breaks in the targeted
library, or a character with the path to a bed file.</p></dd>


<dt id="arg-nontarget">nontarget<a class="anchor" aria-label="anchor" href="#arg-nontarget"></a></dt>
<dd><p>GRanges object with coordinates of breaks in the
non-targeted library, or the path to a bed file. Default: empty GRanges()
object, which is equivalent to not having a control library.</p></dd>


<dt id="arg-region">region<a class="anchor" aria-label="anchor" href="#arg-region"></a></dt>
<dd><p>region around the cutsite to look for alternative breaks.</p></dd>


<dt id="arg-standard-chromosomes">standard_chromosomes<a class="anchor" aria-label="anchor" href="#arg-standard-chromosomes"></a></dt>
<dd><p>logical, constrain the analysis to standard
chromosomes only (defaults to TRUE).</p></dd>


<dt id="arg-bsgenome">bsgenome<a class="anchor" aria-label="anchor" href="#arg-bsgenome"></a></dt>
<dd><p>character, bsgenome to use (eg. BSgenome.Hsapiens.UCSC.hg38).</p></dd>


<dt id="arg-qval-cutoff">qval_cutoff<a class="anchor" aria-label="anchor" href="#arg-qval-cutoff"></a></dt>
<dd><p>numeric, indicating the q-value cutoff, defaults to 0.01.</p></dd>


<dt id="arg-cores">cores<a class="anchor" aria-label="anchor" href="#arg-cores"></a></dt>
<dd><p>Use multiple cores. Defaults to 2 cores, and the function won't
benefit of adding &gt;2. Nevertheless, the speedup with 2 cores is considerable.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>logical, keep informing about every step</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a GRanges object with the coordinates of the cutsite, the number of
breaks on and around the cutsite for the target and non-target libraries,
and a p- and q-values</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The function looks at the number of breaks detected on and around the cutsite
using *only* the breaks detected on the PAM proximal site (looking at reads
on the same strand as the PAM).
Then, a binomial test similar to the one done in the `breakinspectoR()`
function is performed to test for a significant enrichment of breaks around
the cutsite in the target compared to the non-target libraries.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span>  <span class="va">offtargets</span> <span class="op">&lt;-</span> <span class="fu"><a href="breakinspectoR.html">breakinspectoR</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    target   <span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/vegfa.chr6.bed.gz"</span>, package<span class="op">=</span><span class="st">"breakinspectoR"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    nontarget<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/nontarget.chr6.bed.gz"</span>, package<span class="op">=</span><span class="st">"breakinspectoR"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    guide    <span class="op">=</span><span class="st">"GACCCCCTCCACCCCGCCTC"</span>,</span></span>
<span class="r-in"><span>    PAM      <span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"NGG"</span>, <span class="st">"NAG"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    bsgenome <span class="op">=</span><span class="st">"BSgenome.Hsapiens.UCSC.hg38"</span>,</span></span>
<span class="r-in"><span>    eFDR     <span class="op">=</span><span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>    cutsiteFromPAM<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## this is needed only for the package to install</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BSgenome.Hsapiens.UCSC.hg38"</span>, quietly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="co">##</span></span></span>
<span class="r-in"><span>  <span class="va">offtargets.scission_profile</span> <span class="op">&lt;-</span> <span class="fu">scission_profile_analysis</span><span class="op">(</span></span></span>
<span class="r-in"><span>    x        <span class="op">=</span><span class="va">offtargets</span>,</span></span>
<span class="r-in"><span>    target   <span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/vegfa.chr6.bed.gz"</span>, package<span class="op">=</span><span class="st">"breakinspectoR"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    nontarget<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/nontarget.chr6.bed.gz"</span>, package<span class="op">=</span><span class="st">"breakinspectoR"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    bsgenome <span class="op">=</span><span class="st">"BSgenome.Hsapiens.UCSC.hg38"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="fu">mcols</span><span class="op">(</span><span class="va">offtargets</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">offtargets</span><span class="op">)</span>, <span class="fu">mcols</span><span class="op">(</span><span class="va">offtargets.scission_profile</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading BSgenome.Hsapiens.UCSC.hg38...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Importing breaks...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Counting breaks at each offtarget region...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  done</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating probabilities...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  done</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Sergi Sayols.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

